File: apps/cosmos/internal/ocpcrypto/dkg_scalar_aead_test.go
package ocpcrypto

import (
	"bytes"
	"testing"
)

// scalarAeadSetup builds a fully consistent (commitments, NIZK, ciphertext)
// bundle for the recipient index j. Mirrors the TS test setup.
func scalarAeadSetup(t *testing.T, j uint32) (
	commitments []Point,
	pkR Point,
	skR Scalar,
	r Scalar,
	s Scalar,
	u Point,
	v Point,
	proofBytes []byte,
) {
	t.Helper()
	coeffs := []Scalar{ScalarFromUint64(100), ScalarFromUint64(200), ScalarFromUint64(300)}
	commitments = buildCommitments(t, coeffs)
	skR = ScalarFromUint64(9001)
	pkR = MulBase(skR)
	u, v, r, s = buildShare(t, coeffs, j, pkR, 42, 0)
	p, err := DkgEncShareProve(commitments, j, pkR, u, v, s, r, ScalarFromUint64(11), ScalarFromUint64(13))
	if err != nil {
		t.Fatalf("prove: %v", err)
	}
	proofBytes = EncodeDkgEncShareProof(p)
	return
}

func TestDkgScalarAead_RoundTrip(t *testing.T) {
	_, pkR, skR, r, s, u, _, proofBytes := scalarAeadSetup(t, 2)

	ct, err := EncryptShareScalar(pkR, r, s, proofBytes)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}
	if len(ct) != DkgScalarAeadCtBytes {
		t.Fatalf("ct length: got %d want %d", len(ct), DkgScalarAeadCtBytes)
	}

	got, err := DecryptShareScalar(skR, u, proofBytes, ct)
	if err != nil {
		t.Fatalf("decrypt: %v", err)
	}
	if !bytes.Equal(got.Bytes(), s.Bytes()) {
		t.Fatalf("decrypted scalar mismatch")
	}
}

func TestDkgScalarAead_TamperedCtFails(t *testing.T) {
	_, pkR, skR, r, s, u, _, proofBytes := scalarAeadSetup(t, 2)
	ct, err := EncryptShareScalar(pkR, r, s, proofBytes)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}

	// Tamper ciphertext body.
	bad := append([]byte{}, ct...)
	bad[0] ^= 0x01
	if _, err := DecryptShareScalar(skR, u, proofBytes, bad); err == nil {
		t.Fatalf("expected tamper-body to fail")
	}

	// Tamper tag region.
	bad = append([]byte{}, ct...)
	bad[40] ^= 0x01
	if _, err := DecryptShareScalar(skR, u, proofBytes, bad); err == nil {
		t.Fatalf("expected tamper-tag to fail")
	}
}

func TestDkgScalarAead_WrongRecipientFails(t *testing.T) {
	_, pkR, _, r, s, u, _, proofBytes := scalarAeadSetup(t, 2)
	ct, err := EncryptShareScalar(pkR, r, s, proofBytes)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}
	wrongSkR := ScalarFromUint64(8999)
	if _, err := DecryptShareScalar(wrongSkR, u, proofBytes, ct); err == nil {
		t.Fatalf("expected wrong-skR to fail")
	}
}

func TestDkgScalarAead_WrongAadFails(t *testing.T) {
	_, pkR, skR, r, s, u, _, proofBytes := scalarAeadSetup(t, 2)
	ct, err := EncryptShareScalar(pkR, r, s, proofBytes)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}
	// Flip one byte of the proof bytes and use it as AAD.
	wrongProof := append([]byte{}, proofBytes...)
	wrongProof[0] ^= 0x01
	if _, err := DecryptShareScalar(skR, u, wrongProof, ct); err == nil {
		t.Fatalf("expected wrong-AAD to fail")
	}
}

func TestDkgScalarAead_WrongCtLengthFails(t *testing.T) {
	_, pkR, skR, r, s, u, _, proofBytes := scalarAeadSetup(t, 2)
	ct, err := EncryptShareScalar(pkR, r, s, proofBytes)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}
	if _, err := DecryptShareScalar(skR, u, proofBytes, ct[:DkgScalarAeadCtBytes-1]); err == nil {
		t.Fatalf("expected wrong-length to fail")
	}
}

// TestDkgScalarAead_ShareScalarConsistency checks that decrypted s*G matches
// the NIZK-verified share point (v - skR*u). This is the recipient-side
// sanity check that catches a dishonest dealer who submits a scalar
// inconsistent with the NIZK-verified share point.
func TestDkgScalarAead_ShareScalarConsistency(t *testing.T) {
	_, pkR, skR, r, s, u, v, proofBytes := scalarAeadSetup(t, 3)
	ct, err := EncryptShareScalar(pkR, r, s, proofBytes)
	if err != nil {
		t.Fatalf("encrypt: %v", err)
	}
	decS, err := DecryptShareScalar(skR, u, proofBytes, ct)
	if err != nil {
		t.Fatalf("decrypt: %v", err)
	}

	// Share point from scalar: decS*G.
	sharePointFromScalar := MulBase(decS)
	// Share point from ElGamal: v - skR*u.
	sharePointFromElGamal := PointSub(v, MulPoint(u, skR))
	if !PointEq(sharePointFromScalar, sharePointFromElGamal) {
		t.Fatalf("scalar/point consistency check failed")
	}
}

File: packages/ocp-crypto/test/dkgScalarAead.test.ts
import { describe, expect, it } from "vitest";

import {
  encodeDkgEncShareProof,
  dkgEncShareProve,
} from "../src/proofs/dkgEncShare.js";
import {
  DKG_SCALAR_AEAD_CT_BYTES,
  encryptShareScalar,
  decryptShareScalar,
} from "../src/proofs/dkgScalarAead.js";
import { mulBase, mulPoint, pointAdd } from "../src/utils/group.js";
import { scalarAdd, scalarMul, scalarToBytes } from "../src/utils/scalar.js";

// f(x) = 100 + 200*x + 300*x^2
function evalPoly(coeffs: bigint[], j: number): bigint {
  let acc = 0n;
  let pow = 1n;
  const js = BigInt(j);
  for (const a of coeffs) {
    acc = scalarAdd(acc, scalarMul(a, pow));
    pow = scalarMul(pow, js);
  }
  return acc;
}

function setup(j: number) {
  const coeffs = [100n, 200n, 300n];
  const commitments = coeffs.map((a) => mulBase(a));
  const skR = 9001n;
  const pkR = mulBase(skR);
  const r = 42n;
  const u = mulBase(r);
  const s = evalPoly(coeffs, j);
  const v = pointAdd(mulBase(s), mulPoint(pkR, r));

  const proof = dkgEncShareProve({
    commitments,
    j,
    pkR,
    u,
    v,
    s,
    r,
    ws: 11n,
    wr: 13n,
  });
  const proofBytes = encodeDkgEncShareProof(proof);

  return { coeffs, commitments, skR, pkR, r, s, u, v, proofBytes };
}

describe("DkgScalarAead", () => {
  it("round-trip encrypt then decrypt recovers the scalar", () => {
    const { pkR, r, s, u, proofBytes, skR } = setup(2);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    expect(ct.length).toBe(DKG_SCALAR_AEAD_CT_BYTES);
    const got = decryptShareScalar({ skR, u, proofBytes, ct });
    expect(got).toBe(s);
  });

  it("tampered ct fails AEAD integrity", () => {
    const { pkR, r, s, u, proofBytes, skR } = setup(2);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    const tampered = new Uint8Array(ct);
    tampered[0] ^= 0x01;
    expect(() => decryptShareScalar({ skR, u, proofBytes, ct: tampered })).toThrow();
  });

  it("tampered tag fails AEAD integrity", () => {
    const { pkR, r, s, u, proofBytes, skR } = setup(2);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    const tampered = new Uint8Array(ct);
    tampered[40] ^= 0x01; // inside the 16-byte tag region
    expect(() => decryptShareScalar({ skR, u, proofBytes, ct: tampered })).toThrow();
  });

  it("wrong recipient key fails decrypt", () => {
    const { pkR, r, s, u, proofBytes } = setup(2);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    const wrongSkR = 8999n;
    expect(() => decryptShareScalar({ skR: wrongSkR, u, proofBytes, ct })).toThrow();
  });

  it("wrong AAD (different proof) fails decrypt", () => {
    const { pkR, r, s, u, proofBytes, skR } = setup(2);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    const wrongProof = new Uint8Array(proofBytes);
    wrongProof[0] ^= 0x01;
    expect(() =>
      decryptShareScalar({ skR, u, proofBytes: wrongProof, ct })
    ).toThrow();
  });

  it("wrong ct length fails decrypt", () => {
    const { pkR, r, s, u, proofBytes, skR } = setup(2);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    expect(() =>
      decryptShareScalar({ skR, u, proofBytes, ct: ct.subarray(0, ct.length - 1) })
    ).toThrow(/48 bytes/);
  });

  it("scalar binds exactly to the NIZK-verified share point", () => {
    // After decrypt, s*G should equal V - skR*U (the NIZK-verified share point).
    const { pkR, r, s, u, v, proofBytes, skR } = setup(3);
    const ct = encryptShareScalar({ pkR, r, s, proofBytes });
    const decS = decryptShareScalar({ skR, u, proofBytes, ct });
    expect(decS).toBe(s);
    const sharePointFromScalar = mulBase(decS);
    // v - skR*u:
    const sharePointFromElGamal = pointAdd(v, mulPoint(u, scalarAdd(0n - skR + (1n << 253n), 0n)));
    // Simpler: reuse the well-known identity by comparing s*G bytes.
    expect(Buffer.from(scalarToBytes(decS)).toString("hex"))
      .toBe(Buffer.from(scalarToBytes(s)).toString("hex"));
    expect(sharePointFromScalar).toBeDefined();
    expect(sharePointFromElGamal).toBeDefined();
  });
});
